### DISCUSSION TITLE
Can I dynamically import at runtime?

### DISCUSSION THREAD

**Discussion by john-tipper** [vercel/next.js#41309] [Help] [answered]
I'd like to import some javascript resources at runtime via a URL. I'd like to know whether something like this is possible:

```
// perhaps within _document.js
<script type="importmap">
{
  "imports": {
    "mymodule": "https://example.com/mymodule.js",
  }
}
</script>

```

And then within some page, perhaps within `useEffect()`:

```
const my module = await import('mymodule');
```

The above is giving me a compile error, so I don't know if this is due to my syntax, or whether what I want to do is just not possible.  If this is not possible, is there another way by which I can import javascript from a URL at runtime, where the URL is dynamically determined, e.g. on the basis of a header or parameter that is passed in to the request?

**Comment by icyJoseph** [comment, marked as answer]
If I understand your request correctly, I did it like this:

```tsx
import { Html, Head, Main, NextScript } from "next/document";
import Script from "next/script";

export default function Document() {
  return (
    <Html>
      <Head />
      <Script
        id="import-map-script"
        type="importmap"
        dangerouslySetInnerHTML={{
          __html: `  
{
  "imports": {
    "lodash": "https://cdn.jsdelivr.net/npm/lodash@4.17.21/lodash.min.js"
  }
}
    `,
        }}
      ></Script>
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  );
}
```

And then:

```tsx
import "../styles/globals.css";
import type { AppProps } from "next/app";
import { useEffect } from "react";

function MyApp({ Component, pageProps }: AppProps) {
  useEffect(() => {
    // @ts-expect-error
    import("lodash").then((mod) => {
      console.log(mod);
      console.log(mod.default);
      console.log(mod.default.add(10 + 1));
    });
  }, []);

  return <Component {...pageProps} />;
}

export default MyApp;

```

Silencing TypeScript in this case, but I guess you'd need to provide typings for this dynamically loaded modules, which sounds like a lot of work :) 

Tested in development and production builds and works with the regular `script` tag and the `next/script` utility component.

**Reply by john-tipper** [reply]
Thanks, much appreciated.
